### PR TITLE
Implement labeled metrics

### PR DESCRIPTION
**Pull request by badboy**

### CODE REVIEW

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        }
+    }
+
+    /// Destroy this metric.
+    deinit {
+        if self.handle != 0 {
+            switch self.subMetric {
+            case is CounterMetricType:
+                glean_destroy_labeled_counter_metric(self.handle)
+            // TODO(bug XYZ): Requires the BooleanMetricType first
```
Comment (travis79) [inline glean-core/ios/Glean/Metrics/LabeledMetric.swift]: I think you filed a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1583772) for the BooleanMetricType, did you mean to refer to it here?

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            default:
+                // Unreachable. The constructor will already throw an exception on an unhandled sub-metric type
+                assertUnreachable()
+            }
+        }
+    }
+
+    /// Get the specific metric for a given label.
+    ///
+    /// If a set of acceptable labels were specified in the metrics.yaml file,
```
Comment (travis79) [inline glean-core/ios/Glean/Metrics/LabeledMetric.swift]:
```suggestion
    /// If a set of acceptable labels was specified in the metrics.yaml file,
```

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    ///     * label: The label
+    /// - returns: The specific metric for that label
+    subscript(label: String) -> T {
+        // swiftlint:disable force_cast
+        // REASON: We return the same type as the `subMetric` we match against
+
+        switch self.subMetric {
+        case is CounterMetricType:
+            let handle = glean_labeled_counter_metric_get(Glean.shared.handle, self.handle, label)
+            return CounterMetricType(withHandle: handle, disabled: self.disabled, sendInPings: self.sendInPings) as! T
```
Comment (travis79) [inline glean-core/ios/Glean/Metrics/LabeledMetric.swift]: Is there ever a case where the forced cast may fail?

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+/// but records metrics for the underlying metric type `T` in the storage engine for that type.
+/// The only difference is that labeled metrics are stored with the special key `$category.$name/$label`.
+public class LabeledMetricType<T> {
+    let handle: UInt64
+    let disabled: Bool
+    let sendInPings: [String]
+    let labels: Set<String>?
+    let subMetric: T
+
+    /// The public constructor used by automatically generated metrics.
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Metrics/LabeledMetric.swift]: This function throws, would you kindly document the expected exceptions?

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+/// This implements the developer facing API for labeled metrics.
+///
+/// Instances of this class type are automatically generated by the parsers at build time,
+/// allowing developers to record values that were previously registered in the metrics.yaml file.
+///
+/// Unlike most metric types, LabeledMetricType does not have its own corresponding storage engine,
+/// but records metrics for the underlying metric type `T` in the storage engine for that type.
+/// The only difference is that labeled metrics are stored with the special key `$category.$name/$label`.
+public class LabeledMetricType<T> {
```
Comment (Dexterp37) [inline glean-core/ios/Glean/Metrics/LabeledMetric.swift]: Would you kindly add the docs for the labeled metric types, e.g. [`labeled_counter`](https://mozilla.github.io/glean/book/user/metrics/labeled_counters.html)?

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift (hunk 1 above)
Comment (badboy) [reply glean-core/ios/Glean/Metrics/LabeledMetric.swift]: I even changed it locally.... and never committed.

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift (hunk 3 above)
Comment (badboy) [reply glean-core/ios/Glean/Metrics/LabeledMetric.swift]: Nope.
With a simple `as` cast I get an error though:
>  'CounterMetricType' is not convertible to 'T'; did you mean to use 'as!' to force downcast?

File: glean-core/ios/GleanTests/Metrics/LabeledMetricTests.swift [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        )
+
+        labeledStringMetric["label1"].set("foo")
+        labeledStringMetric["label2"].set("bar")
+
+        XCTAssertEqual("foo", try labeledStringMetric["label1"].testGetValue())
+        XCTAssertEqual("bar", try labeledStringMetric["label2"].testGetValue())
+    }
+
+    // TODO(bug XYZ): Requires the BooleanMetricType first
```
Comment (Dexterp37) [inline glean-core/ios/GleanTests/Metrics/LabeledMetricTests.swift]: nit: please update with the bug number here

File: glean-core/ios/Glean/Metrics/LabeledMetric.swift (hunk 3 above)
Comment (travis79) [reply glean-core/ios/Glean/Metrics/LabeledMetric.swift]: If it's guaranteed to safely downcast then `as!` it is! Besides, it is much easier to read than than the safe cast using `as?`.

**Comment by travis79** [comment]
FYI, pending your r+ of #317, you could merge it and go ahead and uncomment the bits that are blocked in this PR if you like.

**Pull request merged**